README.md
[![Build Status](https://travis-ci.org/spheromak/etcd-cookbook.png)](https://travis-ci.org/spheromak/etcd-cookbook)
etcd cookbook - installs coreos/etcd on centos/ubuntu boxen

## Supported Platforms
Centos/rhat 6+ & ubuntu with upstart

## Recipes
| Name | Description |
|:-----|:------------|
| `default` | Install and setup the service
| `binary_install` | Installs the binary of etcd from github release tarballs
| `cluster` | Recipe to aide in the building of multi-node etcd clusters
| `compile_time` | Install and configure etcd during chef compile phase
| `source_install` | Compiles the binary of etcd from a specificed github repo/revision
| `_service` | Recipe used by default for setting up the service

## Attributes

| attribute | default setting | description |
|:---------------------------------|:---------------|:-----------------------------------------|
|`default[:etcd][:install_method]`| `binary` | Right now only binary and source are supported. In the future this will probably go away as there are actual distro packages |
|`default[:etcd][:version]` | `0.3.0` | The release versions to install. binary install will assemble a github url to grab this version |
|`default[:etcd][:sha256]` | `726bd3..` | The Sha256 hash of the tarball specified by the version or URL attribute|
|`default[:etcd][:url]` | `nil` |override the internal generated url to specify your own binary tarball. see the .kitchen.yml for example override |
|`default[:etcd][:state_dir]` | `/var/cache/etcd/state` | Where etcd will store its state |
|`default[:etcd][:search_cook]`| `'etcd\:\:cluster'` | The cookbook that should be searched for on the nodes recipes to detect if it is also running etcd |
|`default[:etcd][:trigger_restart]` | `true` | Make etcd restart if the init config is updated
|`default[:etcd][:trigger_start]` | `true` | Make etcd start during this run
|`default[:etcd][:upstart][:start_on]` | `started networking` | When to start the etcd service using upstart
|`default[:etcd][:upstart][:stop_on]` | `shutdown` | When to stop the etcd service using upstart

These attributes control the startup/cmdline args for etcd:

| attribute | default setting | description |
|:---------------------------------|:---------------|:-----------------------------------------|
|`default[:etcd][:name]`| `nil` | The name etcd uses inthe cluster. By default we use the first available value from `node[:fqdn]`, `node[:hostname]`, or `node[:name]`. Set this to something else to override. |
|`default[:etcd][:snapshot]`| `true` | (deprecated in etcd v2.0 and above) This is really important to get good memory usage. If you're running this in product, you probably want this set to `true` |
|`default[:etcd][:seed_node]` | `nil` | The seed node for initial cluster setup. This node will start as the master, but restarts will rejoin the raft cluster. This needs to be set when using cluster recipe otherwise the cluster will not initialize.|
|`default[:etcd][:nodes]`| `[]` | Array of hostnames in cluster. This provides an alternative method to using Chef's `partial_search` for specifying nodes in cluster (useful if using `chef-solo`)  |
|`default[:etcd][:addr]` | `node.ipaddress:4001` | The address that etcd uses publicly. defaults to  node.ipaddress:4001
|`default[:etcd][:peer_addr]` | `node.ipaddress:7001` | address to announce to peers specified as ip:port or ip when empty string it will set to node.ipaddress:7001
|`default[:etcd][:discovery]` | `''` | Discovery address/token see: https://coreos.com/docs/cluster-management/setup/etcd-cluster-discovery/ for more info
|`default[:etcd][:env_scope]` | `true` | Set the search in cluster recipe to restrict to nodes in the same environment
|`default[:etcd][:args]`| `''` | Extra arguments to pass to etcd when starting the service. if you specify something that is computed they will be passed twice. This is here to handle where you want to setup other things. |

## Usage
#### Default single instance single node:
Simply add etcd to your runlist
````
run_list[etcd]
````

#### Setup a cluster
__NOTE:__ setting `node[:etcd][:addr]` and `[:etcd][:peer_addr]` is important when using a cluster where nodes might have many addresses. Make sure they advertise an address that is reachable by other members.

##### Using Seed node
In a role or wrapper cookbook setup the seed_node attribute and add the cluster recipe to each node in the cluster.
If you use a wrapper cookbook set `node[:etcd][:search_cook]` to the wrapper cookboks name

````
run_list[etcd::cluster]
````

example wrapper can be seen [here](http://github.com/cloudware-cookbooks/ktc-etcd)

##### Using Discovery
To use discovery simply goto [https://discovery.etcd.io/new](https://discovery.etcd.io/new) and get a url then set `node[:etcd][:discovery]` to that url. This can also be your own etcd instance and a key.

Then use `etcd::cluster` recipe to build out the cluster.


## License and Author

|                      |                                                |
|:---------------------|:-----------------------------------------------|
| **Original Author**  | [Jesse Nelson]( https://github.com/spheromak)  |
| **Contributor**      | [Soulou](https://github.com/Soulou)    |
| **Contributor**      | [Aaron O'Mullan](https://github.com/AaronO)    |
| **Contributor**      | [Anthony Scalisi](https://github.com/scalp42)  |
| **Contributor**      | [Robert Coleman](https://github.com/rjocoleman)|
| **Contributor**      | [James Gregory](https://github.com/jagregory)  |
| **Copyright**        | Copyright (c) 2013, Jesse Nelson               |

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
### 2.2.3
* update to etcd 2.0.3

### 2.2.2
* update to etcd 0.4.6
* adds debian support

### 2.2.1
* update to etcd 0.4.5
* Add centos 7 support
* Support centos cloud images without tar
* Dry out compile time recipe
* Move to berks3

### 2.2.0
* update to etcd 0.4.2
* Removed node[:etcd][:local] No longer needed in new etcd. Etcd will default bind :4001 and :7001. 
  The cookbook default is now to use node[:ipaddress] as the  addr and peer_addr.
* add auto-respawn instruction in upstart config file
* Fix Gem deps in the ci build

### 2.1.3
* add peer and peer_addr support / attributes
* add name attribute and computation
* fixup cmdline argument computation
* ensure cluster members are always sorted
* fix spec tests

### 2.1.2
* fix bug in cluster recipe where Resolv was spelled wrong
* fix kitchen test on cluster to pickup this issue.
* add basic chefspec as well to ensure this sort of issue doesn't occur
* fix bug with name attribute methods
* fix searching for wrong recipe when using cluster recipe

### 2.1.1
* update to etcd 0.3.0
* Add support for new discovery mode
* refactor common functions into library methods
* Add specs
* Use rubocop to replace tailor/cane
* Fixup testing (add travis integration) 

### 2.0.0
* update to etcd 0.2.0
* remove `name_switch` attribute. It was marked for deprication in 1.3
* enable snapshotting by default
* add optional local listener
* support chef-solo
* Add support of explicitely specifying a cluster's nodes

### 1.3.5
* silence foodcritic by accessing attributes in a consistent manner, 
* trigger a restart when etcd conf is updated
* include git if source install
* metadata depends on git recipe
* Update Documentation
* Update contributors

### 1.3.4
* fix compile_time to use the right tarball path

### 1.3.3 
* Bump to etcd 1.2
* Added source install recipe

### 1.3.1
* default binding to node[:ipaddress]

### 1.2.4
* update to etcd 0.1.1 release 
* add compile_time recipe that does the whole bit in chef compile time

### 1.2.3
* hotfix to add missing  `node[:etcd][:env_scope]` attribute
* bugfix: use `node[:fqdn]` instead of hostname when matching local machine name

### 1.2.2
* add cluster recipe for setting up clusters of etcd nodes
* re-add seed-node attribute 

### 1.2.1 
* fix the startup args so it is easier to specify custom args
* supports release version and current master on git
* add in tests for both release and git versions of etcd 

### 1.1.1
* Make everything work with release 0.1.0 of etcd
* fixup some syntax issues

### 1.1.0
* move binary install to using coreos/etcd releases from github
* for now the install locattion is fixed to /usr/local and links in /usr/local/bin
* Use ark for managing tarballs
* move bats tests to using etcdctl instead of curl


test/spec/etcd_spec.rb
require_relative 'spec_helper'
require_relative '../../libraries/etcd'

describe 'Etcd' do
  let(:chef_run) { ChefSpec::Runner.new }
  let(:node) { chef_run.node }
  before(:each) do
    Chef::Recipe::Etcd.node = chef_run.node
  end

  context 'package_name' do

    it 'handles version variance' do
      chef_run = ChefSpec::Runner.new
      node = chef_run.node
      Chef::Recipe::Etcd.node = chef_run.node
      node.set[:etcd][:version] = '2.0.3'
      expect(Chef::Recipe::Etcd.package_name).to eql 'etcd-v2.0.3-linux-amd64.tar.gz'
      node.set[:etcd][:version] = '1.3.0'
      expect(Chef::Recipe::Etcd.package_name).to eql 'etcd-v1.3.0-linux-amd64.tar.gz'
      node.set[:etcd][:version] = '0.3.0'
      expect(Chef::Recipe::Etcd.package_name).to eql 'etcd-v0.3.0-linux-amd64.tar.gz'
      node.set[:etcd][:version] = '0.2.0'
      expect(Chef::Recipe::Etcd.package_name).to eql 'etcd-v0.2.0-Linux-x86_64.tar.gz'
      node.set[:etcd][:version] = '0.0.1'
      expect(Chef::Recipe::Etcd.package_name).to eql 'etcd-v0.0.1-Linux-x86_64.tar.gz'
    end
    
    it ' raises when version is nil or invalid' do
      node.set[:etcd][:version] = nil
      expect {Chef::Recipe::Etcd.package_name}.to raise_error
    end
  end

  context 'discovery_cmd' do
    it 'should default to empty string' do
      node.set[:etcd][:discovery] = ''
      node.set[:etcd][:nodes] = []
      Chef::Recipe::Etcd.slave = false
      Chef::Recipe::Etcd.discovery_cmd.should eql ''
    end

    it 'sets up when we have a uuid' do
      node.set[:etcd][:discovery] = '123456'
      Chef::Recipe::Etcd.discovery_cmd.should eql %Q{ -discovery='123456'}
    end

    it 'should setup peers when we are a slave and discovery is disabled.' do
      node.set[:etcd][:discovery] = ''
      Chef::Recipe::Etcd.slave = true
      Chef::Recipe::Etcd.discovery_cmd.should eql ' -peers-file=/etc/etcd_members'
      Chef::Recipe::Etcd.slave = false
    end

  end

  context 'args' do
    before do
      chef_run = ChefSpec::Runner.new
      chef_run.converge('etcd')
      Chef::Recipe::Etcd.node = chef_run.node
    end

    it 'toggles discovery' do
      chef_run = ChefSpec::Runner.new
      chef_run.node.set[:etcd][:discovery] = '1.1.1.1'
      chef_run.converge('etcd')
      Chef::Recipe::Etcd.args.should eql %Q{ -name fauxhai.local -discovery='1.1.1.1' -initial-advertise-peer-urls=10.0.0.2:7001 -advertise-client-urls=10.0.0.2:4001}
    end
  end
end


libraries/etcd.rb
# Encoding: UTF-8
#
# Etcd Helper Libraries
#
#
class Chef
  class Recipe
    # Recipe helpers for etcd cookbook
    module Etcd
      class << self
        attr_accessor :slave, :node

        # return cmd args for discovery/cluster members
        def discovery_cmd
          cmd = ''
          discovery =  node[:etcd][:discovery]
          if discovery.length > 0
            cmd << " -discovery='#{discovery}'"
          elsif slave == true
            cmd << ' -peers-file=/etc/etcd_members'
          end
          cmd
        end

        def lookup_addr(option, key, port)
          cmd = ''
          val = node[:etcd][key.to_sym]
          if val.match(/.*:(\d)/)
            cmd << " #{option}=http://#{val}"
          elsif val.length > 0
            cmd << " #{option}=http://#{val}:#{port}"
          else
            cmd << " #{option}=http://#{node[:ipaddress]}:#{port}"
          end
          cmd
        end

        # determine node name
        def node_name
          a = " -name #{node.name}"
          a = " -name #{node[:fqdn]}" unless node[:fqdn].nil?
          a = " -name #{node[:etcd][:name]}" unless node[:etcd][:name].nil?
          a
        end

        # when you specify args in config we don't compute. so you have to specify all of them
        #
        def args
          cmd = node[:etcd][:args].dup
          cmd << node_name
          cmd << discovery_cmd
          cmd << lookup_addr('-initial-advertise-peer-urls', :initial_advertise_peer_urls, 2380)
          cmd << lookup_addr('-listen-peer-urls', :listen_peer_urls, 2380)
          cmd
        end
        # rubocop:endable MethodLength

        #
        # compute the package name based on etcd version
        #
        def package_name
          version = node[:etcd][:version]
          fail ArgumentError, 'need to specify a version for etcd' unless version
          if Gem::Requirement.new('>= 0.3.0').satisfied_by?(Gem::Version.new(version))
            "etcd-v#{version}-#{node[:os]}-amd64.tar.gz"
          else
            "etcd-v#{version}-#{node[:os].capitalize}-x86_64.tar.gz"
          end
        end

        #
        # Return URL for package via what user has supplied of what we compute
        #
        def bin_url
          version = node[:etcd][:version]
          if  node[:etcd][:url]
            node[:etcd][:url]
          else
            "https://github.com/coreos/etcd/releases/download/v#{version}/#{package_name}"
          end
        end
      end
    end
  end
end


attributes/default.rb
# Encoding: UTF-8
default[:etcd][:install_method] = 'binary'

# address to announce to peers specified as ip:port or ip (will default to :7001)
default[:etcd][:initial_advertise_peer_urls] = ''

# the adress that etcd uses publically. if not set we compute it to node.ipaddress:4001
default[:etcd][:listen_peer_urls] = ''

# set if you want to override the node name. It uses fqdn by default
default[:etcd][:name] = nil

# cluster options
default[:etcd][:seed_node] = nil

# if you wrap this cookbook you should use your wrappers cook name here
default[:etcd][:search_cook] = 'etcd\:\:cluster'

# set to false if you don't want environment scoped searching
default[:etcd][:env_scope] = true

# service start args to pass
default[:etcd][:args] = ''

# v0.3.0 API cluster discovery
default[:etcd][:discovery] = ''

# nodes in cluster
default[:etcd][:nodes] = []

# restart etcd when the config file is updated
default[:etcd][:trigger_restart] = false

# start etcd when etcd is first installed
default[:etcd][:trigger_start] = true

# Upstart parameters for starting/stopping etcd service
default[:etcd][:upstart][:start_on] = 'started networking'
default[:etcd][:upstart][:stop_on] = 'shutdown'

# Release to install
default[:etcd][:version] = '2.0.3'

# Auto respawn
default[:etcd][:respawn] = false

# Sha for github tarball Linux by default
default[:etcd][:sha256] = '0d4dd3ec5c3961433f514544ae7106676f313fe2fa7aa85cde0f2454f1a65b2f'

# Use this to supply your own url to a tarball
default[:etcd][:url] = nil

# Etcd's backend storage
default[:etcd][:state_dir] = '/var/cache/etcd/state'

# Used for source_install method
default[:etcd][:source][:repo] = 'https://github.com/coreos/etcd'
default[:etcd][:source][:revision] = 'HEAD'
default[:etcd][:source][:go_ver] = '1.4.1'
default[:etcd][:source][:go_url] = nil
default[:etcdctl][:source][:repo] = 'https://github.com/coreos/etcdctl'
default[:etcdctl][:source][:revision] = 'HEAD'

# Create user and group for etcd
default[:etcd][:user] = 'etcd'
default[:etcd][:group] = 'etcd'
